trackintel/preprocessing/positionfixes.py

```diff
@@ -173,187 +177,335 @@ def generate_triplegs(positionfixes, staypoints=None, *args, **kwargs):
+        # case 3: Staypoints are not provided but positionfixes have a column 'staypoint_id'
+
+        if staypoints is not None and "staypoint_id" in ret_pfs:
+            case = 1
+        elif staypoints is not None:
+            case = 2
+        elif "staypoint_id" in ret_pfs:
+            case = 3
+        else:
+            raise Exception('unknown case')
 
-        positionfixes_user_this = ret_pfs.loc[ret_pfs['user_id'] == user_id_this]  # this is no copy
-        pfs = positionfixes_user_this.sort_values('tracked_at')
         generated_triplegs = []
+        for user_id_this in ret_pfs['user_id'].unique():
 
-        # Case 1: Staypoints exist and are connected to positionfixes by user id
-        if staypoints is not None and "staypoint_id" in pfs:
-            stps = staypoints.loc[staypoints['user_id'] == user_id_this].sort_values('started_at')
-            stps = stps.reset_index().to_dict('records')
-            for stp1, stp2 in zip(list(stps), list(stps)[1:]):
-                # Get all positionfixes that lie between these two staypoints.
+            positionfixes_user_this = ret_pfs.loc[ret_pfs['user_id'] == user_id_this].sort_values(
+                'tracked_at')  # this is no copy
 
-                # get the last posfix of the first staypoint
-                index_first_posfix_tl = pfs[pfs.staypoint_id == stp1['id']].index[-1]
-                position_first_posfix_tl = pfs.index.get_loc(index_first_posfix_tl)
+            # Case 1: Staypoints exist and are connected to positionfixes by user id
+            if case == 1:
+                generated_triplegs.extend(_triplegs_between_staypoints_case1(positionfixes_user_this, staypoints,
+                                                                             user_id_this))
 
-                # get first posfix of the second staypoint
-                index_last_posfix_tl = pfs[pfs.staypoint_id == stp2['id']].index[0]
-                position_last_posfix_tl = pfs.index.get_loc(index_last_posfix_tl)
+            # Case 2: Staypoints exist but there is no user_id given
+            elif case == 2:
+                generated_triplegs.extend(_triplegs_between_staypoints_case2(positionfixes_user_this, staypoints,
+                                                                             user_id_this))
 
-                pfs_tripleg = pfs.iloc[position_first_posfix_tl:position_last_posfix_tl + 1]
+            # case 3: Only positionfixes with staypoint id for tripleg generation
+            elif case == 3:
+                generated_triplegs.extend(_triplegs_between_staypoints_case3(positionfixes_user_this, user_id_this))
 
-                # include every positionfix that brings you closer to the center 
-                # of the staypoint
+        # create tripleg dataframe
+        if len(generated_triplegs) == 0:
+            ret_tpls = gpd.GeoDataFrame(columns=['id', 'user_id', 'started_at', 'finished_at', 'geom', 'pfs_ids'],
```

Comment:
Here we have 'id' as column, and later also as index. I think the best is to not include `id` to ensure consistency

Reply:
After the if/else is now a ` ret_tpls = ret_tpls.set_index('id')` statement which was missing before. So that should fix that

Reply:
actually you were right and I got confused somehow. I have adjusted it as you said.